src/Reactor.jl

```diff
@@ -13,13 +16,76 @@ struct Reactor{D,Q} <: AbstractReactor
+function Reactor(domain::T,y0::Array{T1,1},tspan::Tuple,interfaces::Z=[];p::X=DiffEqBase.NullParameters(),forwardsensitivities=false,forwarddiff=false,modelingtoolkit=false,tau=1e-3) where {T<:AbstractDomain,T1<:Real,Z<:AbstractArray,X}
     dydt(dy::X,y::T,p::V,t::Q) where {X,T,Q,V} = dydtreactor!(dy,y,t,domain,interfaces,p=p)
     jacy!(J::Q2,y::T,p::V,t::Q) where {Q2,T,Q,V} = jacobiany!(J,y,p,t,domain,interfaces,nothing)
     jacyforwarddiff!(J::Q2,y::T,p::V,t::Q) where {Q2,T,Q,V} = jacobianyforwarddiff!(J,y,p,t,domain,interfaces,nothing)
     jacp!(J::Q2,y::T,p::V,t::Q) where {Q2,T,Q,V} = jacobianp!(J,y,p,t,domain,interfaces,nothing)
     jacpforwarddiff!(J::Q2,y::T,p::V,t::Q) where {Q2,T,Q,V} = jacobianpforwarddiff!(J,y,p,t,domain,interfaces,nothing)
     
+    @inline function _psetup(p::T1, t::T2, u::T3, du::T4, jok::Bool, jcurPtr::T5, gamma::T6, jac!::T7, W::T8, preccache::T9, tau::T10) where {T1,T2,T3,T4,T5,T6,T7,T8,T9,T10}
+        """
+        Update preconditioner when Jacobian needs to be updated
+        p: the parameters
+        t: the current independent variable
+        u: the current state
+        du: the current f(u,p,t)
+        jok: a bool indicating whether the Jacobian needs to be updated
+        jcurPtr: a reference to an Int for whether the Jacobian was updated. jcurPtr[]=true should be set if the Jacobian was updated, and jcurPtr[]=false should be set if the Jacobian was not updated.
+        gamma: the gamma of W = M - gamma*J
+        """
+        if jok
+
+            @. W = 0.0
+            jac!(W,u,p,t)
+            jcurPtr[] = true
+
+            # W = I - gamma*J
+            @. W.nzval = -gamma*W.nzval
+            idxs = diagind(W)
+            @inbounds @views @. W[idxs] = W[idxs] + 1
+
+            # Build preconditioner on W
+            preccache[] = ilu(W, τ = tau)
+        end
+        nothing
+    end
+    
+    @inline function _prec(z::T1, r::T2, p::T3, t::T4, y::T5, fy::T6, gamma::T7, delta::T8, lr::T9, preccache::T10) where {T1,T2,T3,T4,T5,T6,T7,T8,T9,T10}
+        """
+        Compute preccache \\ r in-place and store the result in z
+        z: the computed output vector
+        r: the right-hand side vector of the linear system
+        p: the parameters
+        t: the current independent variable
+        du: the current value of f(u,p,t)
+        gamma: the gamma of W = M - gamma*J
+        delta: the iterative method tolerance
+        lr: a flag for whether lr=1 (left) or lr=2 (right) preconditioning
+        preccache: preconditioner cache
+        """
+        ldiv!(z,preccache[],r)
+    end
+    
+    psetup(p::T1, t::T2, u::T3, du::T4, jok::Bool, jcurPtr::T5, gamma::T6) where {T1,T2,T3,T4,T5,T6} = _psetup(p, t, u, du, jok, jcurPtr, gamma, jacy!, W::SparseMatrixCSC{Float64, Int64}, preccache::Base.RefValue{IncompleteLU.ILUFactorization{Float64, Int64}}, tau::Float64)
```

Comment:
Can we add all the functions necessary to the Reactor object so people can use these with whatever solvers they want, not just our recommended one?

Reply:
Done